rust/src/ima.rs

```diff
@@ -0,0 +1,365 @@
+    variant_utils::new_variant_a_ayay(&map)
+}
+
+struct CommitRewriter<'a> {
+    repo: ostree::Repo,
+    ima: &'a ImaOpts,
+    tempdir: tempfile::TempDir,
+    /// Files that we already changed
+    rewritten_files: HashMap<String, Rc<str>>,
+}
+
+#[context("Gathering xattr {}", k)]
+fn steal_xattr(f: &File, k: &str) -> Result<Vec<u8>> {
+    let k = &CString::new(k)?;
+    unsafe {
+        let k = k.as_ptr() as *const _;
+        let r = libc::fgetxattr(f.as_raw_fd(), k, std::ptr::null_mut(), 0);
+        if r < 0 {
+            return Err(nix::Error::last().into());
+        }
+        let sz: usize = r.try_into()?;
+        let mut buf = vec![0u8; sz];
+        let r = libc::fgetxattr(f.as_raw_fd(), k, buf.as_mut_ptr() as *mut _, sz);
+        if r < 0 {
+            return Err(nix::Error::last().into());
+        }
+        let r = libc::fremovexattr(f.as_raw_fd(), k);
+        if r < 0 {
+            return Err(nix::Error::last().into());
+        }
+        Ok(buf)
+    }
+}
+
+impl<'a> CommitRewriter<'a> {
+    fn new(repo_path: &str, ima: &'a ImaOpts) -> Result<Self> {
+        let repo = ostree::Repo::new_for_path(repo_path);
+        repo.open(gio::NONE_CANCELLABLE)?;
+        Ok(Self {
+            repo,
+            ima,
+            tempdir: tempfile::tempdir_in(Path::new(repo_path).join("tmp"))?,
+            rewritten_files: Default::default(),
+        })
+    }
+
+    /// Use `evmctl` to generate an IMA signature on a file, then
+    /// scrape the xattr value out of it (removing it).
+    ///
+    /// evmctl can write a separate file but it picks the name...so
+    /// we do this hacky dance of `--xattr-user` instead.
+    #[context("Invoking evmctl")]
+    fn ima_sign(
+        &self,
+        instream: &gio::InputStream,
+        selinux: Option<&Vec<u8>>,
+    ) -> Result<HashMap<Vec<u8>, Vec<u8>>> {
+        let mut tempf = tempfile::NamedTempFile::new_in(self.tempdir.path())?;
+        // If we're operating on a bare repo, we can clone the file (copy_file_range) directly.
+        if let Some(instream) = instream.clone().downcast::<gio::UnixInputStream>().ok() {
+            // View the fd as a File
+            let instream_fd = unsafe { File::from_raw_fd(instream.as_raw_fd()) };
+            instream_fd.copy_to(tempf.as_file_mut())?;
+            // Leak to avoid double close
+            let _ = instream_fd.into_raw_fd();
+        } else {
+            // If we're operating on an archive repo, then we need to uncompress
+            // and recompress...
+            let mut instream = instream.clone().into_read();
+            let _n = std::io::copy(&mut instream, tempf.as_file_mut())?;
+        }
+        tempf.seek(std::io::SeekFrom::Start(0))?;
+
+        let mut proc = subprocess::Exec::cmd("evmctl")
+            .args(&[
+                "sign",
+                "--portable",
+                "--xattr-user",
+                "--key",
+                self.ima.key.as_str(),
+            ])
```

Comment:
ah, this is the EVM signature already.

Reply:
I take that this code would create signatures over every file in the compose, is that right?

One more thing about the EVM signatures. Since it is calculated over extended attributes like `security.selinux`, `security.smack64*`, `security.apparmor`, `security.ima`, and `security.capability`, all those have to be available at the time evmctl runs and of course have to have the same value as they will have when the compose is written into a filesystem. Further, for portable EVM signatures the following data from a `stat()` go into the calculation of the hash: `st_uid`, `st_gid`, `st_mode`. So nothing is allowed to manipulate those, no post install script possibly running on the system nothing during boot, no admin, just plainly 'nothing' and 'nobody'. I would try it without EVM first, or at least not activate it, and then if we try it I think it should be more controlled so that config files can be omitted -- at least I would be concerned about those. That's why I liked to do it in the script but EVM cannot be done in a postinstall script...

That said, if you really want to keep it, you should reverse the order, first generate `security.ima`, then `security.evm` at the very end.

Reply:
> Further, for portable EVM signatures the following data from a stat() go into the calculation of the hash: st_uid, st_gid, st_mode. So nothing is allowed to manipulate those, no post install script possibly running on the system nothing during boot, no admin, just plainly 'nothing' and 'nobody'.

Yes, that's actually *exactly* what we want for ostree.  We *do* want the ability to make hardlinks - that's how the whole transactional updates work.  But otherwise ostree ensure that content that it hardlinks goes underneath a read-only bind mount (mostly in `/usr`) and indeed we absolutely under no circumstances want anything to mutate those because that corrupts the object store.

Reply:
> Yes, that's actually _exactly_ what we want for ostree. We _do_ want the ability to make hardlinks - that's how the whole transactional updates work. But otherwise ostree ensure that content that it hardlinks goes underneath a read-only bind mount (mostly in `/usr`) and indeed we absolutely under no circumstances want anything to mutate those because that corrupts the object store.

As long as you filter out/skip anything that's legally mutable, such as config files.

Reply:
Right, ostree has both `/usr/etc` (immutable) and `/etc`.  In this case actually, ostree may need to learn to strip off the IMA xattrs when copying files into `/etc`.